plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java

```diff
@@ -60,73 +63,82 @@
         Map<String, Processor.Factory> processorFactories = new HashMap<>();
         for (Map.Entry<String, ProcessorFactoryProvider> entry : processorFactoryProviders.entrySet()) {
             Processor.Factory processorFactory = entry.getValue().get(environment);
             processorFactories.put(entry.getKey(), processorFactory);
         }
         this.processorFactoryRegistry = Collections.unmodifiableMap(processorFactories);
-        clusterService.add(new PipelineStoreListener());
     }
 
     @Override
-    protected void doStart() {
-    }
-
-    @Override
-    protected void doStop() {
-    }
-
-    @Override
-    protected void doClose() {
-        // TODO: When org.elasticsearch.node.Node can close Closable instances we should remove this code
+    public void close() throws IOException {
+        // TODO: When org.elasticsearch.node.Node can close Closable instances we should try to remove this code,
+        // since any wired closable should be able to close itself
         List<Closeable> closeables = new ArrayList<>();
         for (Processor.Factory factory : processorFactoryRegistry.values()) {
             if (factory instanceof Closeable) {
                 closeables.add((Closeable) factory);
             }
         }
-        try {
-            IOUtils.close(closeables);
-        } catch (IOException e) {
-            throw new RuntimeException(e);
-        }
+        IOUtils.close(closeables);
     }
 
     /**
      * Deletes the pipeline specified by id in the request.
      */
     public void delete(DeletePipelineRequest request, ActionListener<DeleteResponse> listener) {
+        ensureReady();
+
         DeleteRequest deleteRequest = new DeleteRequest(request);
         deleteRequest.index(PipelineStore.INDEX);
         deleteRequest.type(PipelineStore.TYPE);
         deleteRequest.id(request.id());
         deleteRequest.refresh(true);
-        client().delete(deleteRequest, listener);
+        client().delete(deleteRequest, new ActionListener<DeleteResponse>() {
+            @Override
+            public void onResponse(DeleteResponse deleteResponse) {
+                client().execute(RefreshAction.INSTANCE, new RefreshRequest(), new ActionListener<RefreshResponse>() {
+                    @Override
+                    public void onResponse(RefreshResponse refreshResponse) {
+                        // TODO: do we want to include in the response how many ingest nodes successfully refreshed their pipelines?
+                        listener.onResponse(deleteResponse);
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        logger.error("Failed to update pipelines on other ingest nodes [{}]", e);
+                        // TODO: do we want to fail the request here?
+                        // We did manage to successfully store the pipeline document, but ingest nodes where unable to load
```

Comment:
s/where/were  in the comment.

I am not sure what we should do here. Change is stored but in order for it to be seen we need to wait for the next refresh in the background. Failing might sound like a good idea, but then people can't do anything about it, cause the refresh api is internal and can't be called through REST.